src/v/cloud_storage/cache_service.cc

```diff
@@ -1288,42 +1289,55 @@ ss::future<> cache::put(
     options.io_priority_class = io_priority;
     auto out = co_await ss::make_file_output_stream(tmp_cache_file, options);
 
-    std::exception_ptr disk_full_error;
+    std::exception_ptr eptr;
+    bool no_space_on_device = false;
     try {
         co_await ss::copy(data, out)
           .then([&out]() { return out.flush(); })
           .finally([&out]() { return out.close(); });
     } catch (const std::filesystem::filesystem_error& e) {
         // For ENOSPC errors, delay handling so that we can do a trim
-        if (e.code() == std::errc::no_space_on_device) {
-            disk_full_error = std::current_exception();
-        } else {
-            throw;
-        }
+        no_space_on_device = e.code() == std::errc::no_space_on_device;
+        eptr = std::current_exception();
+    } catch (...) {
+        // For other errors, delay handling so that we can clean up the tmp file
+        eptr = std::current_exception();
     }
 
-    if (disk_full_error) {
-        vlog(cst_log.error, "Out of space while writing to cache");
+    // If we failed to write to the tmp file, we should delete it, maybe do an
+    // eager trim, and rethrow the exception.
+    if (eptr) {
+        try {
+            co_await delete_file_and_empty_parents(tmp_filepath.native());
```

Comment:
This should probably be done only if the `eptr` is not a shutdown exception. No need to delay shutdown.

Reply:
I don't agree with this. `ssx::is_shutdown_exception` shouldn't be used here as it can't disambiguate between an abort caused requested by application logic i.e. kafka request being aborted vs application being shut down.

Although I couldn't find any users with this logic interacting with cache service there can be in the future. We also often throw `ss::abort_requested_exception` deliberately in a few places during normal operations.

Reply:
I commented on is_shutdown_exception previously as well https://github.com/redpanda-data/redpanda/pull/14946#discussion_r1392466938. Imho we should avoid it as a plague. It is an excuse for bad design/code.

Reply:
I can understand the stance in general, but I'm confused why not in this case? It seems the only shutdown error we can get in delete_file_and_empty_parents is gate closed when the cache is being shut down. Am I missing something?

Reply:
I believe @Lazin is referring to eptr which comes from ss::copy() and pulling from the input stream rather than `delete_file_and_empty_parents`. Aka don't run cleanup if we're shutting down. But my point is that we can't distinguish if the exception we got is because e.g. http client is shutting down or someone aborted the connection because the caller isn't interested in the result anymore.

Reply:
The input stream which is used here is not http input stream. It actually comes from `cloud_io::io_resources` and it can throw shutdown exception when redpanda is shutting down. I don't think that anything in redpanda should throw `broken_promise` during normal operation. If something does this not during the shutdown it indicates a bug. The problem with invoking `delete_file_and_empty_parents` during shutdown is that it will likely throw a shutdown exception again. There are several scheduling points there so when you will get the shutdown error it's quite possible that the `cache` is also shutting down. The `delete_file_and_empty_parents` is holding the gate so it will throw `gate_closed` exception right away.

Reply:
Shortcutting some code during shutdown is not a matter of performance (shutting down quickly is important though). A lot of code will barf if invoked during the shutdown. This creates a lot of corner cases which are not tested and are quite difficult to test. In this case when you're getting `broker_sempahore` from the token bucket inside the stream this means that the cache itself may or may not be shutting down at the moment. So basically we have two corner cases here only for the shutdown case.